suite/tests/client-interface/drstatecmp-fuzz-app.c

```diff
@@ -0,0 +1,218 @@
+
+static uint32_t
+rand_32b(void)
+{
+    uint32_t r = rand() & 0xff;
+    r |= (rand() & 0xff) << 8;
+    r |= (rand() & 0xff) << 16;
+    r |= (rand() & 0xff) << 24;
+    return r;
+}
+
+static uint32_t
+generate_encoded_inst(void)
+{
+    /* Pick one of the available (side-effect-free and non-branch) opcodes
+     * and randomize the non-fixed bits.
+     */
+    uint32_t opcode_pick = rand() % FUZZ_INST_CNT;
+    const opcode_opnd_pair_t opcode_opnd_pair = fuzz_opcode_opnd_pairs[opcode_pick];
+
+    uint32_t opnd_mask = opcode_opnd_pair.opnd;
+    /* Avoid registers 16-31 for Rd (special registers and callee-saved registers). */
+    opnd_mask &= 0b01111;
+    /* Fuzz the operand bits. */
+    uint32_t fuzzed_opnd = rand_32b() & opnd_mask;
+
+    uint32_t encoded_inst = opcode_opnd_pair.opcode | fuzzed_opnd;
+    return encoded_inst;
+}
+
+static int
+check_decoded_inst(instr_t *decoded_inst)
+{
+    return instr_valid(decoded_inst) && instr_get_opcode(decoded_inst) != OP_xx &&
+        instr_raw_bits_valid(decoded_inst) && instr_operands_valid(decoded_inst);
+}
+
+static byte *
+generate_inst(byte *encode_pc)
+{
+    /* Pick a random side-effect-free and non-branch instruction. */
+    uint32_t encoded_inst = generate_encoded_inst();
+    byte encoded_inst_bytes[4];
+    for (int i = 0; i < 4; ++i, encoded_inst >>= 8)
+        encoded_inst_bytes[i] = (byte)(encoded_inst & 0xff);
+
+    /* Try to decode the randomized encoding. */
+    instr_t *decoded_inst = instr_create(GLOBAL_DCONTEXT);
+    byte *nxt_pc = decode(GLOBAL_DCONTEXT, encoded_inst_bytes, decoded_inst);
+    /* XXX: Ideally the decoder would report as erroneous any encoding leading to SIGILL.
+     * Currently, several valid decodings are illegal instructions.
+     */
+    if (nxt_pc != NULL && check_decoded_inst(decoded_inst))
+        encode_pc = append_instr(decoded_inst, encode_pc);
+    else
+        instr_destroy(GLOBAL_DCONTEXT, decoded_inst);
+
+    return encode_pc;
+}
+
+static void
+generate_code()
+{
+    /* Account for the generated insts and the final return. */
+    code_size = (NUM_INSTS + 1) * 4;
+    generated_code =
+        (byte *)allocate_mem(code_size, ALLOW_EXEC | ALLOW_READ | ALLOW_WRITE);
+    assert(generated_code != NULL);
+
+    /* Synthesize code which includes a lot of side-effect-free instructions. Only one
+     * basic block is created (linear control flow). To test clobbering of arithmetic
+     * flags conditionally-executed instructions are included.
+     */
+    byte *encode_pc = generated_code;
+    for (int i = 0; i < NUM_INSTS; ++i) {
+        encode_pc = generate_inst(encode_pc);
+    }
+
+    /* The outer level is a function. */
+    encode_pc = append_instr(XINST_CREATE_return(GLOBAL_DCONTEXT), encode_pc);
+    assert(encode_pc && encode_pc <= generated_code + code_size);
+    protect_mem(generated_code, code_size, ALLOW_EXEC | ALLOW_READ);
+}
+
+int
+main(void)
+{
+    /* Produce fuzzing application code. */
+    srand(time(NULL));
```

Comment:
Add an XXX comment about taking in the seed as a parameter, and printing it out on an error, to reproduce the exact behavior in a subsequent run.

Reply:
Done.